main/ts_client.c

```diff
@@ -5,67 +5,251 @@
+        // hackky solution so that free() can always be called on names_string
+        char *msg = "No devices Connected";
+        names_string = (char *) malloc(strlen(msg)+1*sizeof(char));
+        strncpy(names_string, msg, strlen(msg)+1);
+    }
+    return names_string;
+}
+TSDevice *ts_get_device(char *device_id)
+{
+    int i = 0;
+    while (devices[i] != NULL) {
+        if (strcmp(devices[i]->ts_device_id, device_id) == 0) {
+            return devices[i];
         }
+        i++;
     }
     return NULL;
 }
 
-int ts_resp_status(char *resp)
+// wrapper for strlen() to check for NULL
+static int strlen_null(char *r)
 {
-    unsigned int status_code = 0;
-    int ret = sscanf(resp, ":%X ", &status_code);
-    if (ret > 0) {
-        return status_code;
+    if (r != NULL) {
+        return(strlen(r));
+    } else {
+        return 0;
     }
-    else {
-        return -1;
+}
+
+char *exec_or_create(char *node)
+{
+    if (strstr(node, "auth") != NULL || strstr(node, "exec") != NULL) {
+        return "!";
+    } else {
+        return "+";
     }
 }
 
-int ts_req_hdr_from_http(char *buf, size_t buf_size, int method, const char *uri)
+void ts_parse_uri(const char *uri, TSUriElems *params)
 {
-    ESP_LOGI(TAG, "URI: %s, method: %d", uri, method);
+    if (uri == NULL || !strcmp(uri, "")) {
+        ESP_LOGE(TAG, "Got invalid uri");
+        return;
+    }
+    params->ts_list_subnodes = uri[strlen(uri)] == '/' ? 0 : 1;
 
-    switch (method) {
-        case HTTP_GET:
-            buf[0] = '?';
+    // copy uri so we can safely modify
+    char *temp_uri = (char *) heap_caps_malloc(sizeof(char)*(strlen(uri)+1), MALLOC_CAP_8BIT);
+    if (temp_uri == NULL) {
+        ESP_LOGE(TAG, "Unable to allocate memory for temp_uri");
+        return;
+    }
+    strcpy(temp_uri, uri);
+    // extract device ID
+    int i = 0;
+    while (temp_uri[i] != '\0' && temp_uri[i] != '/') {
+        i++;
+    }
+    // replace '/' so we have two null-terminated strings
+    temp_uri[i] = '\0';
+    params->ts_device_id = temp_uri;
+    // this points either to '\0' aka NULL or the rest of the string
+    params->ts_target_node = temp_uri+i+1;
+    ESP_LOGD(TAG, "Device_id: %s", params->ts_device_id);
+    ESP_LOGD(TAG, "Target Node: %s", params->ts_target_node);
+    ESP_LOGD(TAG, "List the sub nodes: %s", params->ts_list_subnodes == 1 ? "yes" : "no");
+}
+
+char *ts_build_query(uint8_t ts_method, TSUriElems *params)
+{
+    // calculate size to allocate buffer
+    int nbytes = 3;  // method + termination + zero termination
+    if (*(params->ts_target_node) == '\0' && params->ts_list_subnodes == 1) {
+        nbytes++;    // '/' at the end of query
+    }
+    nbytes += strlen_null(params->ts_target_node);
+    if (params->ts_payload != NULL) {
+        nbytes += strlen_null(params->ts_payload) +1;    // additional whitespace between uri and array/json
+    }
+    char *ts_query = (char *) heap_caps_malloc(sizeof(char)*nbytes, MALLOC_CAP_8BIT);
```

Comment:
A `char` is always 1 byte long.
```suggestion
    char *ts_query = (char *) heap_caps_malloc(nbytes, MALLOC_CAP_8BIT);
```

Reply:
Yes, I just got used to always use `sizeof()` for better portability